Map screenshot and map scale widget

matthewma7 opened the pull request:
Implement taking map screenshot feature with a map ui layer widget

![2018-03-05_17-54-24](https://user-images.githubusercontent.com/3123478/37043666-3af92f44-212f-11e8-9e98-8ecb7d514b85.gif)

manthey commented on web_external/views/map/MapPanel.js:
```diff
@@ -418,6 +430,20 @@ const MapPanel = Panel.extend({
+        screenshotButtonContainer.empty().append(screenshotButtonTemplate());
+        screenshotButtonContainer.find('button').on('click', () => {
+            var layers = this.map.layers().filter((layer) => layer.layerName !== 'ui');
+            this.map.screenshot(layers, { wait: 'idle' }).then((image) => {
```
The `map.screenshot` function should either be `map.screenshot(layers, undefined, undefined, {wait: 'idle'})` or `map.screenshot({layers: layers, wait: 'idle'})`.

aashish24 commented:
@matthewma7 is this ready for review? Looks like travis is failing.

matthewma7 commented:
@aashish24 David and I looked into why it's failing. David is fixing the issue in Girder and GeoJS. I'll change this to [WIP]

manthey commented:
Fundamentally, there is a [webkit bug](https://bugs.webkit.org/show_bug.cgi?id=17352) (also see [here](https://bugs.webkit.org/show_bug.cgi?id=29305) and [here](https://bugs.webkit.org/show_bug.cgi?id=129172))  that prevents webkit-based browsers from using the canvas to take a screenshot of HTML components.  This applies to PhantomJS (which we use for testing), iOS, and probably Safari.

There is a minor issue in the GeoJS screenshot code in that if a css file is missing, the screenshot fails.  I'll have a PR for that soon.  But this doesn't fix the underlying webit issue.  For webkit browsers, the best we can do with a screenshot is get the canvas elements (webGL and 2d canvas layers), and skip the html and svg layers (though some svg would be possible with a bunch of work).

I think the best thing to do would be the first time a screenshot is attempted that wants an HTML layer, we test if such a thing is supported.  If not, we emit a console warning and then skip rendering such layers.  This has the virtue that webkit browsers would still get screenshots of base maps and many features (and, thus, not be completely broken).  The Minerva screenshot test would succeed (though would not contain the scale in the PhantomJS test).  Unless there is some reason not to, I'll make another GeoJS PR to this effect.

matthewma7 commented:
Updated to latest GeoJS and fixed the build error. Thanks @manthey 

Also, added scale widget to the map.
![image](https://user-images.githubusercontent.com/3123478/38741245-825f7fdc-3f07-11e8-9501-8d3c13954a6c.png)

manthey commented on plugin_tests/client/minervaSpec.js:
```diff
@@ -42,7 +42,7 @@ describe('Main view', function () {
                     layer.createWidget = function (widgetName) {
                         if (widgetName === 'slider') {
-                            return {};
+                            return { canvas: function () { } };
```
At least on the latest GeoJS, I don't think we need this `beforeEach` and `afterEach`.  At least, locally, the tests all pass without these.

manthey commented on web_external/views/map/MapPanel.js:
```diff
@@ -130,13 +138,25 @@ const MapPanel = Panel.extend({
                     : {});
             this.uiLayer = this.map.createLayer('ui');
-            this.uiLayer.createWidget('slider', { position: { right: 20, bottom: 30 } });
+            this.uiLayer2 = this.map.createLayer('ui');
```
This would be clearer with a more explicit name (`controlLayer` perhaps?) and a comment saying that widgets on this layer won't be included in the screenshot.

manthey commented:
@matthewma7 This is failing due to pip 10.0's release.  I'll make a PR to address that soon.
